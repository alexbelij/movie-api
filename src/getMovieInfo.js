/* @flow */

import R from 'ramda';

import connector from './kinopoisk/connector';

const kinopoiskCdnUrl = 'http://st.kp.yandex.net/images';
const normalizeCdnImageUrl = (imageUrl: string) =>
  `${kinopoiskCdnUrl}/${imageUrl}`
    .replace('iphone_', 'iphone360_')
    .replace('iphone60', 'iphone360');

type KinopoiskApi$GetFilmResponse = {
  filmID?: ?string,
  imdbID?: ?string,
  nameRU?: ?string,
  nameEN?: ?string,
  posterURL?: ?string,
  year?: ?string,
  description?: ?string,
  filmLength?: ?string,
  genre?: ?string,
  ratingAgeLimits?: ?string,
  ratingMPAA?: ?string,
  ratingData: {
    rating?: ?string,
    ratingVoteCount?: ?string,
    ratingIMDb?: ?string,
    ratingIMDbVoteCount?: ?string,
    ratingFilmCritics?: ?string,
    ratingFilmCriticsVoteCount?: ?string,
  },
};

const parseRuntime = (rawRuntime: ?string): number => (
  rawRuntime
    ? R.pipe(
        R.split(':'),
        R.slice(0, 2),
        ([hours, minutes]: [string, string]): number => R.sum([
          parseInt(hours, 10) * 60,
          parseInt(minutes, 10),
        ]),
        (runtime: number) => (runtime === 0 ? NaN : runtime),
      )(rawRuntime)
    : NaN
);

const getInfo = async (id: number) => {
  const res: ?KinopoiskApi$GetFilmResponse =
    await connector.get('getFilm', { filmID: id });

  if (!res) return null;

  const {
    filmID,
    imdbID,
    nameRU,
    nameEN,
    posterURL,
    year,
    description,
    filmLength,
    genre,
    ratingAgeLimits,
    ratingMPAA,
    ratingData = {},
  } = res;

  return {
    kpId: parseInt(filmID, 10),
    imdbID,
    title: {
      ru: nameRU,
      original: nameEN,
    },
    posterUrl: posterURL ? normalizeCdnImageUrl(posterURL) : null,
    year: parseInt(year, 10),
    synopsis: {
      ru: description,
    },
    runtime: parseRuntime(filmLength),
    genres: (genre || '').split(', '),
    ageRating: parseInt(ratingAgeLimits, 10),
    mpaaRating: ratingMPAA,
    kpRating: parseFloat(ratingData.rating),
    kpRatingVoteCount: parseInt(ratingData.ratingVoteCount, 10),
    imdbRating: parseFloat(ratingData.ratingIMDb),
    imdbRatingVoteCount: parseInt(ratingData.ratingIMDbVoteCount, 10),
    rtCriticsRating: parseInt(ratingData.ratingFilmCritics, 10),
    rtCriticsRatingVoteCount: parseInt(
      ratingData.ratingFilmCriticsVoteCount,
      10,
    ),
  };
};

type KinopoiskApi$StaffMember = {
  id: string,
  type: 'KPPeople',
  nameRU?: ?string,
  nameEN?: ?string,
  posterURL?: ?string,
  description?: ?string,
  professionText: string,
  professionKey: 'actor' | 'director' | 'producer' | 'writer' | 'composer',
};
type KinopoiskApi$GetStaffResponse = {
  creators?: Array<Array<KinopoiskApi$StaffMember>>,
};
type StaffMember = {
  name: {
    ru: ?string,
    en: ?string,
  },
  photoUrl: ?string,
};

const getCrew = async (id: number): Promise<?{
  cast: Array<StaffMember>,
  directors: Array<StaffMember>,
  producers: Array<StaffMember>,
  writers: Array<StaffMember>,
  composers: Array<StaffMember>,
}> => {
  const res: ?KinopoiskApi$GetStaffResponse =
    await connector.get('getStaff', { filmID: id });

  if (!res) return null;

  return R.pipe(
    R.flatten,
    R.groupBy(R.prop('professionKey')),
    R.pick(['actor', 'director', 'producer', 'writer', 'composer']),
    R.map(R.pipe(
      R.slice(0, 10),
      R.map(({ nameRU, nameEN, posterURL }: KinopoiskApi$StaffMember) => ({
        name: {
          ru: nameRU,
          en: nameEN,
        },
        photoUrl: posterURL ? normalizeCdnImageUrl(posterURL) : null,
      })),
    )),
    ({ actor, director, producer, writer, composer }: Object) => ({
      cast: actor,
      directors: director,
      producers: producer,
      writers: writer,
      composers: composer,
    }),
  )(res.creators);
};

type KinopoiskApi$GalleryItem = {
  image: string,
  preview: string,
};
type KinopoiskApi$GetGalleryResponse = {
  gallery?: {
    kadr?: ?Array<KinopoiskApi$GalleryItem>,
    kadr_sp?: ?Array<KinopoiskApi$GalleryItem>,
    poster?: ?Array<KinopoiskApi$GalleryItem>,
  },
};

const getStills = async (id: number): Promise<?Array<string>> => {
  const res: ?KinopoiskApi$GetGalleryResponse =
    await connector.get('getGallery', { filmID: id });

  if (!res) return null;

  return R.pipe(
    R.propOr([], 'kadr'),
    R.map(
      ({ image }: KinopoiskApi$GalleryItem) => normalizeCdnImageUrl(image),
    ),
  )(res.gallery);
};

const getMovieInfo = async (id: number) => {
  const [info, crew, stills] = await Promise.all([
    getInfo(id),
    getCrew(id),
    getStills(id),
  ]);

  if (!info) return null;

  return { ...info, ...(crew || {}), stills: stills || [] };
};

export {
  getInfo as __getInfo,
  getCrew as __getCrew,
  getStills as __getStills,
};
export default getMovieInfo;
