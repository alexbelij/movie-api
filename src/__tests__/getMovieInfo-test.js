/* @flow */

import { __getInfo, __getCrew, __getStills } from '../getMovieInfo';

const kpId = 714888; // Star Wars: The Force Awakens

describe('getMovieInfo', () => {
  it('fetches movie info from kinopoisk for a given id', async () => {
    expect(await __getInfo(kpId)).toMatchSnapshot();
  });

  it('fetches movie crew from kinopoisk for a given id', async () => {
    expect(await __getCrew(kpId)).toMatchSnapshot();
  });

  it('fetches movie stills from kinopoisk for a given id', async () => {
    expect(await __getStills(kpId)).toMatchSnapshot();
  });
});
