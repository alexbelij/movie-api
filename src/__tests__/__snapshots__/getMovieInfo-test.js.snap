exports[`getMovieInfo fetches movie crew from kinopoisk for a given id 1`] = `
Object {
  "cast": Array [
    Object {
      "name": Object {
        "en": "John Boyega",
        "ru": "Джон Бойега",
      },
      "photoUrl": "http://st.kp.yandex.net/images/actor_iphone/iphone360_2196854.jpg?d=20130703131657",
    },
    Object {
      "name": Object {
        "en": "Daisy Ridley",
        "ru": "Дэйзи Ридли",
      },
      "photoUrl": "http://st.kp.yandex.net/images/actor_iphone/iphone360_3016071.jpg?d=20140430113652",
    },
    Object {
      "name": Object {
        "en": "Oscar Isaac",
        "ru": "Оскар Айзек",
      },
      "photoUrl": "http://st.kp.yandex.net/images/actor_iphone/iphone360_43284.jpg",
    },
    Object {
      "name": Object {
        "en": "Adam Driver",
        "ru": "Адам Драйвер",
      },
      "photoUrl": "http://st.kp.yandex.net/images/actor_iphone/iphone360_1833677.jpg?d=20130521201357",
    },
    Object {
      "name": Object {
        "en": "Domhnall Gleeson",
        "ru": "Донал Глисон",
      },
      "photoUrl": "http://st.kp.yandex.net/images/actor_iphone/iphone360_671252.jpg?d=20150622214531",
    },
    Object {
      "name": Object {
        "en": "Harrison Ford",
        "ru": "Харрисон Форд",
      },
      "photoUrl": "http://st.kp.yandex.net/images/actor_iphone/iphone360_5679.jpg?d=20130506141058",
    },
    Object {
      "name": Object {
        "en": "Mark Hamill",
        "ru": "Марк Хэмилл",
      },
      "photoUrl": "http://st.kp.yandex.net/images/actor_iphone/iphone360_23716.jpg?d=20130705122542",
    },
    Object {
      "name": Object {
        "en": "Carrie Fisher",
        "ru": "Кэрри Фишер",
      },
      "photoUrl": "http://st.kp.yandex.net/images/actor_iphone/iphone360_8459.jpg?d=20130705123958",
    },
    Object {
      "name": Object {
        "en": "Max von Sydow",
        "ru": "Макс фон Сюдов",
      },
      "photoUrl": "http://st.kp.yandex.net/images/actor_iphone/iphone360_3311.jpg",
    },
    Object {
      "name": Object {
        "en": "Andy Serkis",
        "ru": "Энди Серкис",
      },
      "photoUrl": "http://st.kp.yandex.net/images/actor_iphone/iphone360_13176.jpg",
    },
  ],
  "composers": Array [
    Object {
      "name": Object {
        "en": "John Williams",
        "ru": "Джон Уильямс",
      },
      "photoUrl": "http://st.kp.yandex.net/images/actor_iphone/iphone360_225027.jpg?d=20130908235640",
    },
  ],
  "directors": Array [
    Object {
      "name": Object {
        "en": "J.J. Abrams",
        "ru": "Джей Джей Абрамс",
      },
      "photoUrl": "http://st.kp.yandex.net/images/actor_iphone/iphone360_27741.jpg?d=20150827134539",
    },
  ],
  "producers": Array [
    Object {
      "name": Object {
        "en": "J.J. Abrams",
        "ru": "Джей Джей Абрамс",
      },
      "photoUrl": "http://st.kp.yandex.net/images/actor_iphone/iphone360_27741.jpg?d=20150827134539",
    },
    Object {
      "name": Object {
        "en": "Bryan Burk",
        "ru": "Брайан Берк",
      },
      "photoUrl": "http://st.kp.yandex.net/images/actor_iphone/iphone360_663746.jpg",
    },
    Object {
      "name": Object {
        "en": "Kathleen Kennedy",
        "ru": "Кэтлин Кеннеди",
      },
      "photoUrl": "http://st.kp.yandex.net/images/actor_iphone/iphone360_20396.jpg?d=20150721001100",
    },
    Object {
      "name": Object {
        "en": "Pippa Anderson",
        "ru": "Пиппа Андерсон",
      },
      "photoUrl": null,
    },
    Object {
      "name": Object {
        "en": "Michael Arndt",
        "ru": "Майкл Арндт",
      },
      "photoUrl": "http://st.kp.yandex.net/images/actor_iphone/iphone360_936886.jpg?d=20130612003448",
    },
    Object {
      "name": Object {
        "en": "Leifur B. Dagfinnsson",
        "ru": "Лейфур Б. Дагфиннссон",
      },
      "photoUrl": null,
    },
    Object {
      "name": Object {
        "en": "Tommy Gormley",
        "ru": "Томми Гормли",
      },
      "photoUrl": null,
    },
    Object {
      "name": Object {
        "en": "Tommy Harper",
        "ru": "Томми Харпер",
      },
      "photoUrl": null,
    },
    Object {
      "name": Object {
        "en": "Jason D. McGatlin",
        "ru": "Джейсон Д. МакГатлин",
      },
      "photoUrl": "http://st.kp.yandex.net/images/actor_iphone/iphone360_43725.jpg",
    },
    Object {
      "name": Object {
        "en": "Michelle Rejwan",
        "ru": "Мишель Рейван",
      },
      "photoUrl": null,
    },
  ],
  "writers": Array [
    Object {
      "name": Object {
        "en": "Lawrence Kasdan",
        "ru": "Лоуренс Кэздан",
      },
      "photoUrl": "http://st.kp.yandex.net/images/actor_iphone/iphone360_26537.jpg?d=20131017121035",
    },
    Object {
      "name": Object {
        "en": "J.J. Abrams",
        "ru": "Джей Джей Абрамс",
      },
      "photoUrl": "http://st.kp.yandex.net/images/actor_iphone/iphone360_27741.jpg?d=20150827134539",
    },
    Object {
      "name": Object {
        "en": "Michael Arndt",
        "ru": "Майкл Арндт",
      },
      "photoUrl": "http://st.kp.yandex.net/images/actor_iphone/iphone360_936886.jpg?d=20130612003448",
    },
    Object {
      "name": Object {
        "en": "George Lucas",
        "ru": "Джордж Лукас",
      },
      "photoUrl": "http://st.kp.yandex.net/images/actor_iphone/iphone360_24284.jpg",
    },
  ],
}
`;

exports[`getMovieInfo fetches movie info from kinopoisk for a given id 1`] = `
Object {
  "ageRating": 12,
  "genres": Array [
    "фантастика",
    "фэнтези",
    "боевик",
  ],
  "imdbID": "tt5613052",
  "imdbRating": 8.2,
  "imdbRatingVoteCount": 590,
  "kpId": 714888,
  "kpRating": 7.2,
  "kpRatingVoteCount": 113,
  "mpaaRating": "PG-13",
  "posterUrl": "http://st.kp.yandex.net/images/film_iphone/iphone360_714888.jpg?d=20151105180111",
  "rtCriticsRating": 92,
  "rtCriticsRatingVoteCount": 346,
  "runtime": 138,
  "synopsis": Object {
    "ru": "Через тридцать лет после гибели Дарта Вейдера и Императора галактика по-прежнему в опасности. Государственное образование Первый Орден во главе с таинственным верховным лидером Сноуком и его правой рукой Кайло Реном идёт по стопам Империи, пытаясь захватить всю власть. В это нелёгкое время судьба сводит юную девушку Рей и бывшего штурмовика Первого Ордена Финна с героями войны с Империей - Ханом Соло, Чубаккой и генералом Леей. Вместе они должны дать бой Первому Ордену, однако настаёт тот момент, когда становится очевидно, что лишь джедаи могут остановить Сноука и Кайло Рена.",
  },
  "title": Object {
    "original": "Star Wars: Episode VII - The Force Awakens",
    "ru": "Звёздные войны: Пробуждение силы",
  },
  "year": 2015,
}
`;

exports[`getMovieInfo fetches movie stills from kinopoisk for a given id 1`] = `
Array [
  "http://st.kp.yandex.net/images/kadr/2751407.jpg",
  "http://st.kp.yandex.net/images/kadr/2751406.jpg",
  "http://st.kp.yandex.net/images/kadr/2751405.jpg",
  "http://st.kp.yandex.net/images/kadr/2751404.jpg",
  "http://st.kp.yandex.net/images/kadr/2751403.jpg",
  "http://st.kp.yandex.net/images/kadr/2751402.jpg",
  "http://st.kp.yandex.net/images/kadr/2751401.jpg",
  "http://st.kp.yandex.net/images/kadr/2751400.jpg",
  "http://st.kp.yandex.net/images/kadr/2751399.jpg",
  "http://st.kp.yandex.net/images/kadr/2751390.jpg",
  "http://st.kp.yandex.net/images/kadr/2751389.jpg",
  "http://st.kp.yandex.net/images/kadr/2751388.jpg",
  "http://st.kp.yandex.net/images/kadr/2735779.jpg",
  "http://st.kp.yandex.net/images/kadr/2735778.jpg",
  "http://st.kp.yandex.net/images/kadr/2735777.jpg",
  "http://st.kp.yandex.net/images/kadr/2735776.jpg",
  "http://st.kp.yandex.net/images/kadr/2735775.jpg",
  "http://st.kp.yandex.net/images/kadr/2735774.jpg",
  "http://st.kp.yandex.net/images/kadr/2735773.jpg",
  "http://st.kp.yandex.net/images/kadr/2735772.jpg",
  "http://st.kp.yandex.net/images/kadr/2735771.jpg",
  "http://st.kp.yandex.net/images/kadr/2735770.jpg",
  "http://st.kp.yandex.net/images/kadr/2735769.jpg",
  "http://st.kp.yandex.net/images/kadr/2735768.jpg",
  "http://st.kp.yandex.net/images/kadr/2735767.jpg",
  "http://st.kp.yandex.net/images/kadr/2735766.jpg",
  "http://st.kp.yandex.net/images/kadr/2735765.jpg",
  "http://st.kp.yandex.net/images/kadr/2735764.jpg",
  "http://st.kp.yandex.net/images/kadr/2735763.jpg",
  "http://st.kp.yandex.net/images/kadr/2735762.jpg",
  "http://st.kp.yandex.net/images/kadr/2704437.jpg",
  "http://st.kp.yandex.net/images/kadr/2704436.jpg",
  "http://st.kp.yandex.net/images/kadr/2704435.jpg",
  "http://st.kp.yandex.net/images/kadr/2704434.jpg",
  "http://st.kp.yandex.net/images/kadr/2704433.jpg",
  "http://st.kp.yandex.net/images/kadr/2704432.jpg",
  "http://st.kp.yandex.net/images/kadr/2704431.jpg",
  "http://st.kp.yandex.net/images/kadr/2704430.jpg",
  "http://st.kp.yandex.net/images/kadr/2696992.jpg",
  "http://st.kp.yandex.net/images/kadr/2696371.jpg",
  "http://st.kp.yandex.net/images/kadr/2696370.jpg",
  "http://st.kp.yandex.net/images/kadr/2696369.jpg",
  "http://st.kp.yandex.net/images/kadr/2696368.jpg",
  "http://st.kp.yandex.net/images/kadr/2696367.jpg",
  "http://st.kp.yandex.net/images/kadr/2696366.jpg",
  "http://st.kp.yandex.net/images/kadr/2696365.jpg",
  "http://st.kp.yandex.net/images/kadr/2696364.jpg",
  "http://st.kp.yandex.net/images/kadr/2694487.jpg",
  "http://st.kp.yandex.net/images/kadr/2693086.jpg",
  "http://st.kp.yandex.net/images/kadr/2693085.jpg",
  "http://st.kp.yandex.net/images/kadr/2693084.jpg",
  "http://st.kp.yandex.net/images/kadr/2693083.jpg",
  "http://st.kp.yandex.net/images/kadr/2693082.jpg",
  "http://st.kp.yandex.net/images/kadr/2693080.jpg",
  "http://st.kp.yandex.net/images/kadr/2689711.jpg",
  "http://st.kp.yandex.net/images/kadr/2689496.jpg",
  "http://st.kp.yandex.net/images/kadr/2689492.jpg",
  "http://st.kp.yandex.net/images/kadr/2687963.jpg",
  "http://st.kp.yandex.net/images/kadr/2687962.jpg",
  "http://st.kp.yandex.net/images/kadr/2687961.jpg",
  "http://st.kp.yandex.net/images/kadr/2687960.jpg",
  "http://st.kp.yandex.net/images/kadr/2687959.jpg",
  "http://st.kp.yandex.net/images/kadr/2687958.jpg",
  "http://st.kp.yandex.net/images/kadr/2687957.jpg",
  "http://st.kp.yandex.net/images/kadr/2687956.jpg",
  "http://st.kp.yandex.net/images/kadr/2687955.jpg",
  "http://st.kp.yandex.net/images/kadr/2687954.jpg",
  "http://st.kp.yandex.net/images/kadr/2687953.jpg",
  "http://st.kp.yandex.net/images/kadr/2687952.jpg",
  "http://st.kp.yandex.net/images/kadr/2687951.jpg",
  "http://st.kp.yandex.net/images/kadr/2687950.jpg",
  "http://st.kp.yandex.net/images/kadr/2687949.jpg",
  "http://st.kp.yandex.net/images/kadr/2687948.jpg",
  "http://st.kp.yandex.net/images/kadr/2687947.jpg",
  "http://st.kp.yandex.net/images/kadr/2687946.jpg",
  "http://st.kp.yandex.net/images/kadr/2687945.jpg",
  "http://st.kp.yandex.net/images/kadr/2687944.jpg",
  "http://st.kp.yandex.net/images/kadr/2687943.jpg",
  "http://st.kp.yandex.net/images/kadr/2687942.jpg",
  "http://st.kp.yandex.net/images/kadr/2687941.jpg",
  "http://st.kp.yandex.net/images/kadr/2687940.jpg",
  "http://st.kp.yandex.net/images/kadr/2687939.jpg",
  "http://st.kp.yandex.net/images/kadr/2687938.jpg",
  "http://st.kp.yandex.net/images/kadr/2687937.jpg",
  "http://st.kp.yandex.net/images/kadr/2687936.jpg",
  "http://st.kp.yandex.net/images/kadr/2683178.jpg",
  "http://st.kp.yandex.net/images/kadr/2681409.jpg",
  "http://st.kp.yandex.net/images/kadr/2665335.jpg",
  "http://st.kp.yandex.net/images/kadr/2665334.jpg",
  "http://st.kp.yandex.net/images/kadr/2665333.jpg",
  "http://st.kp.yandex.net/images/kadr/2665332.jpg",
  "http://st.kp.yandex.net/images/kadr/2665331.jpg",
  "http://st.kp.yandex.net/images/kadr/2665330.jpg",
  "http://st.kp.yandex.net/images/kadr/2665329.jpg",
  "http://st.kp.yandex.net/images/kadr/2665328.jpg",
  "http://st.kp.yandex.net/images/kadr/2665327.jpg",
  "http://st.kp.yandex.net/images/kadr/2665326.jpg",
  "http://st.kp.yandex.net/images/kadr/2665325.jpg",
  "http://st.kp.yandex.net/images/kadr/2665324.jpg",
  "http://st.kp.yandex.net/images/kadr/2665323.jpg",
  "http://st.kp.yandex.net/images/kadr/2665322.jpg",
  "http://st.kp.yandex.net/images/kadr/2665321.jpg",
  "http://st.kp.yandex.net/images/kadr/2664388.jpg",
  "http://st.kp.yandex.net/images/kadr/2646409.jpg",
  "http://st.kp.yandex.net/images/kadr/2646408.jpg",
  "http://st.kp.yandex.net/images/kadr/2631723.jpg",
  "http://st.kp.yandex.net/images/kadr/2631722.jpg",
  "http://st.kp.yandex.net/images/kadr/2631721.jpg",
  "http://st.kp.yandex.net/images/kadr/2631720.jpg",
  "http://st.kp.yandex.net/images/kadr/2631719.jpg",
  "http://st.kp.yandex.net/images/kadr/2631718.jpg",
  "http://st.kp.yandex.net/images/kadr/2631717.jpg",
  "http://st.kp.yandex.net/images/kadr/2631716.jpg",
  "http://st.kp.yandex.net/images/kadr/2631715.jpg",
  "http://st.kp.yandex.net/images/kadr/2601907.jpg",
  "http://st.kp.yandex.net/images/kadr/2583585.jpg",
  "http://st.kp.yandex.net/images/kadr/2583584.jpg",
  "http://st.kp.yandex.net/images/kadr/2578266.jpg",
  "http://st.kp.yandex.net/images/kadr/2578265.jpg",
  "http://st.kp.yandex.net/images/kadr/2578264.jpg",
  "http://st.kp.yandex.net/images/kadr/2578263.jpg",
  "http://st.kp.yandex.net/images/kadr/2578262.jpg",
  "http://st.kp.yandex.net/images/kadr/2578261.jpg",
  "http://st.kp.yandex.net/images/kadr/2578260.jpg",
  "http://st.kp.yandex.net/images/kadr/2578259.jpg",
  "http://st.kp.yandex.net/images/kadr/2578258.jpg",
  "http://st.kp.yandex.net/images/kadr/2578257.jpg",
  "http://st.kp.yandex.net/images/kadr/2578256.jpg",
  "http://st.kp.yandex.net/images/kadr/2578255.jpg",
  "http://st.kp.yandex.net/images/kadr/2578254.jpg",
  "http://st.kp.yandex.net/images/kadr/2578253.jpg",
  "http://st.kp.yandex.net/images/kadr/2578252.jpg",
  "http://st.kp.yandex.net/images/kadr/2530193.jpg",
  "http://st.kp.yandex.net/images/kadr/2517360.jpg",
  "http://st.kp.yandex.net/images/kadr/2517359.jpg",
  "http://st.kp.yandex.net/images/kadr/2517358.jpg",
  "http://st.kp.yandex.net/images/kadr/2517357.jpg",
]
`;
